### ISSUE TITLE
ruff.toml ignored inside subdir

### ISSUE THREAD

**Issue by ctcjab** [python-lsp/python-lsp-ruff#102]
Reproduction Steps
1. `uv venv && . .venv/bin/activate`
2. `uv pip install jupyterlab jupyterlab-lsp python-lsp-ruff ruff` [1]
3. `jupyter lab --notebook-dir=/tmp`
4. Create a top-level notebook with the code `import os` and confirm that it triggers error F401, as exected
5. Create a top-level ruff.toml with `ignore = ["F401"]` in the `[lint]` section
6. Reload the notebook from disk and confirm that error F401 is no longer triggered, as expected
7. Close the notebook
8. Move both the notebook and the ruff.toml into any subdirectory
9. Open the notebook from inside the subdir
10. Observe that the ruff.toml is ignored and error F401 is triggered again
11. The same bug occurs when using pyproject.toml instead of (or in addition to) ruff.toml

Screen capture of running through this repro:
https://github.com/user-attachments/assets/8cb0710a-0ebf-4279-8633-ab9a13a39901

<details><summary>[1] Click to see `pip freeze` output</summary>

```
❯ uv pip freeze
Using Python 3.12.9 environment at: jup-venv
anyio==4.8.0
appnope==0.1.4
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
asttokens==3.0.0
async-lru==2.0.4
attrs==25.1.0
babel==2.17.0
beautifulsoup4==4.13.3
bleach==6.2.0
cattrs==24.1.2
certifi==2025.1.31
cffi==1.17.1
charset-normalizer==3.4.1
comm==0.2.2
debugpy==1.8.13
decorator==5.2.1
defusedxml==0.7.1
docstring-to-markdown==0.15
executing==2.2.0
fastjsonschema==2.21.1
fqdn==1.5.1
h11==0.14.0
httpcore==1.0.7
httpx==0.28.1
idna==3.10
ipykernel==6.29.5
ipython==9.0.1
ipython-pygments-lexers==1.1.1
isoduration==20.11.0
jedi==0.19.2
jinja2==3.1.6
json5==0.10.0
jsonpointer==3.0.0
jsonschema==4.23.0
jsonschema-specifications==2024.10.1
jupyter-client==8.6.3
jupyter-core==5.7.2
jupyter-events==0.12.0
jupyter-lsp==2.2.5
jupyter-server==2.15.0
jupyter-server-terminals==0.5.3
jupyterlab==4.3.5
jupyterlab-lsp==5.1.0
jupyterlab-pygments==0.3.0
jupyterlab-server==2.27.3
lsprotocol==2023.0.1
markupsafe==3.0.2
matplotlib-inline==0.1.7
mistune==3.1.2
nbclient==0.10.2
nbconvert==7.16.6
nbformat==5.10.4
nest-asyncio==1.6.0
notebook-shim==0.2.4
overrides==7.7.0
packaging==24.2
pandocfilters==1.5.1
parso==0.8.4
pexpect==4.9.0
platformdirs==4.3.6
pluggy==1.5.0
prometheus-client==0.21.1
prompt-toolkit==3.0.50
psutil==7.0.0
ptyprocess==0.7.0
pure-eval==0.2.3
pycparser==2.22
pygments==2.19.1
python-dateutil==2.9.0.post0
python-json-logger==3.3.0
python-lsp-jsonrpc==1.1.2
python-lsp-ruff==2.2.2
python-lsp-server==1.12.2
pyyaml==6.0.2
pyzmq==26.2.1
referencing==0.36.2
requests==2.32.3
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rpds-py==0.23.1
ruff==0.9.10
send2trash==1.8.3
setuptools==75.8.2
six==1.17.0
sniffio==1.3.1
soupsieve==2.6
stack-data==0.6.3
terminado==0.18.1
tinycss2==1.4.0
tornado==6.4.2
traitlets==5.14.3
types-python-dateutil==2.9.0.20241206
typing-extensions==4.12.2
ujson==5.10.0
uri-template==1.3.0
urllib3==2.3.0
wcwidth==0.2.13
webcolors==24.11.1
webencodings==0.5.1
websocket-client==1.8.0
```
</details>

**Comment by jhossbach** [comment]
Can you increase the verbosity and send the output? See https://github.com/python-lsp/python-lsp-ruff?tab=readme-ov-file#debugging

**Comment by jhossbach** [comment]
Actually, I misunderstood the problem. `python-lsp-ruff` does not actually load the configuration from the `ruff.toml`, it only looks for it to prevent pylsp from accidentally overwriting the project settings with the lsp settings: https://github.com/python-lsp/python-lsp-ruff/blob/041c695d829239345ee9dd5c1f017311d6725ff0/pylsp_ruff/plugin.py#L743-L745

So this is not an issue of `pylsp` or `pylsp-ruff`, but rather `ruff` not finding the `ruff.toml`. Could it be that the CWD in `jupyterlab` is set to the parent directory and `ruff` therefore cannot see the `ruff.toml`?
I am not sure if this is something that should be changed in the first place, since reading a configuration file in a subdirectory below the CWD would probably break some configs for people?
Also not sure who to blame here, either ruff or jupyterlab-lsp?

**Comment by ctcjab** [comment]
> Could it be that the CWD in jupyterlab is set to the parent directory and ruff therefore cannot see the ruff.toml?

If that were the case, I would not expect this to work, but it does:
```console
❯ cat subdir/foo.py 
import os

❯ cat subdir/ruff.toml 
[lint]
ignore = ["F401"]

❯ ruff check subdir/foo.py 
All checks passed!

❯ mv subdir/ruff.toml{,.disabled} 

❯ ruff check subdir/foo.py 
subdir/foo.py:1:8: F401 [*] `os` imported but unused
  |
1 | import os
  |        ^^ F401
  |
  = help: Remove unused import: `os`

Found 1 error.
[*] 1 fixable with the `--fix` option.
```
I.e. ruff itself respects ruff.toml files in subdirectories of the CWD when checking adjacent (or descendent) sources.

> Can you increase the verbosity and send the output? See https://github.com/python-lsp/python-lsp-ruff?tab=readme-ov-file#debugging

I tried for a long time and still could not figure out how to do this. :(

Thanks for any help looking into this.

**Comment by jhossbach** [comment]
`python-lsp-ruff` parses the python file to ruff via stdin and gives the relative path to the file via `--stdin-filename` for ruff to find the correct configuration file. I did some more digging and found that here the notebook in the virtual documents directory is used, therefore ignoring the `ruff.toml` in the subdirectory. In other words:
```bash
cat <...>| python -m ruff check --stdin-filename=subdir/Untitled.ipynb --extension=ipynb:python
```
instead of 
```bash
<cat ...>| python -m ruff check --stdin-filename=.virtual_documents/subdir/Untitled.ipynb --extension=ipynb:python
```
I think this can be achieved by removing the virtual documents dir from the filename: https://jupyterlab-lsp.readthedocs.io/en/latest/Configuring.html#virtual-documents-dir
I'll look into it.

**Comment by jhossbach** [comment]
I did some playing around. The issue can be solved trivially if `JP_LSP_VIRTUAL_DIR` is not set to a custom directory. Otherwise `pylsp` never knows the real path, only the path to the virtual document. In this case, I am not sure how to handle this from `pylsp` or `pylsp-ruff`'s side.
I will publish the changes soon.

**Comment by jhossbach** [comment]
Assuming you have not set the `JP_LSP_VIRTUAL_DIR` env variable, you could place a `ruff.toml` with the required config into `.virtual_documents`. Soft-linking should also work.
I don't think the problem is fixable from pylsps side AFAIK, since the LSP-server never gets information about the true CWD, only the virtual directory and its contents

One solution would be to copy/link the `ruff.toml` to the virtual directory, such that when calling ruff with `--stdin-filename` pointing to the virtual directory, ruff will look inside that directory for configurations.
Or to give pylsp the path to the actual directory. Not sure what is easier to implement.
I will open up an issue on jupyterlab-lsp soon (or you can do it)

**Comment by ctcjab** [comment]
Hi @jhossbach, thanks for the response and sorry for the delay.

> The issue can be solved trivially if JP_LSP_VIRTUAL_DIR is not set to a custom directory.

I can confirm `JP_LSP_VIRTUAL_DIR` is unset (including when I perform the reproduction steps I listed above), so I don't think this is due to setting `JP_LSP_VIRTUAL_DIR` to a custom directory, so it's not solvable by ceasing to do so.

> Assuming you have not set the JP_LSP_VIRTUAL_DIR env variable, you could place a ruff.toml with the required config into .virtual_documents. Soft-linking should also work.

Unfortunately using `.virtual_documents/ruff.toml` is not going to work here, for a few reasons. Primarily, we need to preserve the invariant that `ruff` always uses the ruff.toml maintained in version control alongside the code that `ruff` is linting. This way, CI's pass/fail result is a pure function of the contents of the repo at the revision under test. (Creating a ruff.toml symlink in `.virtual_documents` doesn't work either as there are multiple clones that users work with, each of which has a ruff.toml that needs to be honored.)

> I will open up an issue on jupyterlab-lsp soon (or you can do it)

Would you mind opening the issue, since I think you understand the problem better? (I searched https://github.com/jupyter-lsp/jupyterlab-lsp/issues but don't see one there yet.) Thank you very much for your help with this.

**Comment by ctcjab** [comment]
(As an aside, I hit [this very nasty bug](https://github.com/jupyter-lsp/jupyterlab-lsp/issues/1129) while testing this and lost a lot of work 😿.)

**Comment by jhossbach** [comment]
Hey, sorry for the delay.
> Unfortunately using .virtual_documents/ruff.toml is not going to work here, for a few reasons.

I am not sure I understand. There should be a single `ruff.toml` that everyone is using, and placing this toml into the .virtual_documents should at least make it visible to pylsp.
Do you mean there are ruff.toml files in subdirectories of that directory that need to be taken into account for jupyter notebooks in that directory as well?

**Comment by jhossbach** [comment]
I can create the issue, but I am still not sure what the fix would be. Does copying the ruff.toml file into the .virtual_documents directory actually fix the problem? When you copy that file into .virtual_documents does it solve your issue?

**Comment by ctcjab** [comment]
https://github.com/jupyter-lsp/jupyterlab-lsp/pull/1168 has now been merged (and hopefully will be released soon).

That just leaves the (1.) still TODO from my previous comment. @jhossbach, might you (or someone you can ping here) be able to take a look when you get a chance? Thank you 🙏

**Comment by ctcjab** [comment]
Hi @jhossbach, still hitting this, and noticed https://github.com/jupyter-lsp/jupyterlab-lsp/pull/1168 was released in v5.3.0 on April 2. Just thought I'd check back in here since it's been another month, hope you don't mind.

**Comment by jhossbach** [comment]
Yes, it is still on mind. The problem is on the Jupyterlab-LSP side, if nothing new happens there I will go ahead and submit a PR there with the required changes and see how it goes.
I will tackle it some time this or next week

**Comment by jhossbach** [comment]
@ctcjab sorry for the wait, could you switch to `fix/virtual_documents_dir` and see if the changes resolve the issue?

**Comment by jhossbach** [comment]
Hey @ctcjab, did you have any chance to check out this other branch?
